doc/designs/mlkem.md

```diff
@@ -0,0 +1,91 @@
+functions associated with each set.
+
+To support these variants, OpenSSL will have 3 different key managers and 3
+corresponding KEM function sets. The names used are of the form "ML-KEM-768".
+
+The key problem implementing the different variants lies in the significant
+differences in (dimensions of) vectors and matrics required. For this reason,
+[boringssl](https://boringssl.googlesource.com/boringssl/+/HEAD/crypto/mlkem)
+chose to use C++ templates to represent the different parameter sets.
+As C++ cannot be used in OpenSSL, (object) code duplication by way of use
+of macros to minimize source code duplication should be used.
+
+Substantial logic for the algorithm(s) is provided via SHA3 code. To improve
+execution performance the required objects are pre-allocated and maintained
+via an MLKEM_CTX object.
+
+This MLKEM_CTX is then passed to all functions. A convenience function exists
+to properly create it. It also is automatically allocated in the providers'
+MLKEM key context but a more efficient placement would be within the
+provider context (TODO(ML-KEM)).
+
+ML-KEM keys
+-----------
+
+ML-KEM keys are separated as usual into public and private keys. The public
+key format is rather well-defined and can get exported/imported via proper
+marshaling functions. The private key format is somewhat more contested still:
+FIPS 203 documents a format that is commonly referred to as the "extended"
+format and also exportable/importable via marshaling functions in well-defined
+sizes. The IETF voices interest in using the "seed-based" format that is
+supported by way of functions dubbed "internal and for testing only".
+
+The design therefore caters for both options: The default key generation and
+KEM encapsulation/decapsulation functions operate on/with "extended keys".
+It is also possible to use the "seed-based" format by way of providing
+specific OSSL_PARAMs made available for that purpose -- but again, as per
+NIST guidance, only for testing. If the seed version is retrieved from a
+normal key generation operation, it shall be subject to the same level of
+protection given to private key material.
+
+Key generation API
+------------------
+
+Keys can therefore be generated as "usual" by way of the EVP functions
+EVP_PKEY_generate() and EVP_PKEY_Q_keygen().
+
+The "seed format" can be selected by either setting the OSSL_PARAM value
+"OSSL_PKEY_PARAM_MLKEM_SEED" before key generation or retrieving it
+after key generation.
+
+KEM API
+-------
+
+ML-KEM is meant to be a simple replacement for existing KEM algorithms.
+Therefore, simple use should be
+
+EVP_PKEY_encapsulate_init(), EVP_PKEY_encapsulate(),
+EVP_PKEY_decapsulate_init(), EVP_PKEY_decapsulate().
+
+For the encapsulation operation, a test-only option exists to avoid the
+otherwise mandatory use of a random number generator for passing in a
+known "entropy" by way of the OSSL_PARAM "OSSL_KEM_PARAM_MLKEM_ENC_ENTROPY".
```

Comment:
Over in #25990 I'm changing this parameter to `OSSL_KEM_PARAM_IKME` (_"ikme"_) so it aligns with the similar option used by the ECX algorithms.

It's not clear if it would be better to change it here too (before either merges).

Reply:
Thanks for the head-up.

> It's not clear if it would be better to change it here too (before either merges).

My proposal would be to adapt this PR to whatever #25990 winds up encoding and merge it only at the end (when consensus has built what makes most sense/is most coherent with the pre-existing code base).

Reply:
Thanks for the heads-up (and letting me know about #25990 :)!

> It's not clear if it would be better to change it here too (before either merges).

My thinking was to adapt this PR to whatever #26054 winds up encoding and merge it only at the end (when consensus has built what makes most sense/is most coherent with the new and the pre-existing code base). But I'd also be fine closing this PR if either you or @vdukhovni also want to add the documentation to #26054 straight.

Reply:
I've pulled my changes out into #26082 which is likely to merge quickly.
I also fixed some bugs in the new evp_test KEM code.

Reply:
FYI @paulidale we decided to merge this pretty much as-is to close out a sprint and then change docs again as new code lands.

Reply:
OK, we'll need to remember to do that. The parameter name will in fact soon be: `OSSL_KEM_PARAM_IKME`.
Here, I'd have been tempted to just get it over with, and replace the name with that.